File: examples/C/server_example.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include "rmt_server.h"

static char *my_interface = NULL;
static char interface[50];

char *short_options = "n:h";
struct option long_options[] = {
    {"net",  required_argument, NULL, 'n'},
    {"help", no_argument,       NULL, 'h'},
    { 0,     0,                 0,    0  },
};

void print_help(void)
{
    printf("Usage: ./server_example [options] [server_cmd]\n");
    printf("options:\n");
    printf("  * --help: Showing this messages.\n");
    printf("  * --net [interface]: Decide which interface agent uses.\n");
    printf("server_cmd:\n");
    printf("  * search: show the search result (default).\n");
    printf("  * set: set the config from certain ID.\n");
    printf("  * task-get: get the task list and current task mode from certain ID.\n");
    printf("  * task-set: set the task mode to certain ID.\n");
    printf("  * sendfile: send files to agent.\n");
    printf("  * recvfile: receive files from agent.\n");
    printf("  * all: do all the action of search, get, set.\n");
}

typedef enum _SVR_CMD {
    CMD_SEARCH = 0,
    CMD_SET,
    CMD_GET_TASK,
    CMD_SET_TASK,
    CMD_SEND_FILE,
    CMD_RECV_FILE,
    CMD_ALL,
    CMD_SUM
} SVR_CMD;

char *svr_cmd_mapping[CMD_SUM] = {
    "search",
    "set",
    "task-get",
    "task-set",
    "sendfile",
    "recvfile",
    "all"
};

void server_cmd_search_and_get(void)
{
    device_info *dev_ptr;
    int dev_num;
    unsigned long *id_list;
    data_info *info_list;
    int info_list_num;

    // get device list
    dev_ptr = rmt_server_create_device_list(&dev_num);
    for (int i = 0; i < dev_num; i++) {
        printf("Device %d\n", i);
        printf("ID: %lu\n", dev_ptr[i].deviceID);
        printf("Model: %s\n", dev_ptr[i].model);
        printf("Host: %s\n", dev_ptr[i].host);
        printf("IP: %s\n", dev_ptr[i].ip);
        printf("MAC: %s\n", dev_ptr[i].mac);
        printf("RMT version: %s\n", dev_ptr[i].rmt_version);
        printf("Device Info: %s\n", dev_ptr[i].devinfo);
        fflush (stdout);
    }
    // assign id to id_list
    id_list = (unsigned long *) malloc(sizeof(unsigned long) * dev_num);
    for (int i = 0; i < dev_num; i++) {
        id_list[i] = dev_ptr[i].deviceID;
    }
    // get server_info
    info_list = rmt_server_get_info(id_list, dev_num, "cpu;ram;hostname;wifi;", &info_list_num);
    printf("Try to get info from %d device\n", info_list_num);
    for (int i = 0; i < info_list_num; i++) {
        printf("ID: %ld\n", info_list[i].deviceID);
        printf("return list: %s\n", info_list[i].value_list);
    }
    rmt_server_free_info(info_list);
    free(id_list);

    rmt_server_free_device_list(dev_ptr);
}

void server_cmd_set(void)
{
    data_info set_info;
    data_info *info_list;
    int info_list_num;

    // set data_info
    printf("Try to set info to id 6166\n");
    set_info.deviceID = 6166;
    strncpy(set_info.value_list, "hostname:ros-ROScube-I;locate:on;", CONFIG_KEY_STR_LEN);
    info_list = rmt_server_set_info(&set_info, 1, &info_list_num);
    for (int i = 0; i < info_list_num; i++) {
        printf("ID: %ld\n", info_list[i].deviceID);
        printf("return list: %s\n", info_list[i].value_list);
    }
    rmt_server_free_info(info_list);
    sleep(2);

    // set same info
    unsigned long set_id_list[1] = {6166};
    printf("Try to set info to id 6166 with same value\n");
    info_list = rmt_server_set_info_with_same_value(set_id_list, 1, "hostname:ros-ROScube-I;locate:off", &info_list_num);
    for (int i = 0; i < info_list_num; i++) {
        printf("ID: %ld\n", info_list[i].deviceID);
        printf("return list: %s\n", info_list[i].value_list);
    }
}

void server_cmd_set_task(void)
{
    data_info set_info;
    data_info *info_list;
    int info_list_num;

    // set task to Navigation
    printf("Try to set task to id 6166\n");
    set_info.deviceID = 6166;
    strncpy(set_info.value_list, "task_mode:Navigation", CONFIG_KEY_STR_LEN);
    info_list = rmt_server_set_info(&set_info, 1, &info_list_num);
    for (int i = 0; i < info_list_num; i++) {
        printf("ID: %ld\n", info_list[i].deviceID);
        printf("return list: %s\n", info_list[i].value_list);
    }

#if 0 // open it if you wnat to test 'Idle' task mode
    sleep(5);

    // set task to Idle
    printf("Try to set task to id 6166\n");
    set_info.deviceID = 6166;
    strncpy(set_info.value_list, "task_mode:Idle", CONFIG_KEY_STR_LEN);
    info_list = rmt_server_set_info(&set_info, 1, &info_list_num);
    for (int i = 0; i < info_list_num; i++) {
        printf("ID: %ld\n", info_list[i].deviceID);
        printf("return list: %s\n", info_list[i].value_list);
    }
#endif

    rmt_server_free_info(info_list);
}

void server_cmd_get_task(void)
{
    device_info *dev_ptr;
    int dev_num;
    unsigned long *id_list;
    data_info *info_list;
    int info_list_num;

    // get device list
    dev_ptr = rmt_server_create_device_list(&dev_num);

    // assign id to id_list
    id_list = (unsigned long *) malloc(sizeof(unsigned long) * dev_num);
    for (int i = 0; i < dev_num; i++) {
        id_list[i] = dev_ptr[i].deviceID;
    }

    // get task info
    info_list = rmt_server_get_info(id_list, dev_num, "task_list;task_mode", &info_list_num);
    printf("Try to get task info from %d device\n", info_list_num);
    for (int i = 0; i < info_list_num; i++) {
        printf("ID: %ld\n", info_list[i].deviceID);
        printf("return list: %s\n", info_list[i].value_list);
    }
    rmt_server_free_info(info_list);
    free(id_list);

    rmt_server_free_device_list(dev_ptr);
}

void server_cmd_send_file(void)
{
    transfer_result file_result;
    transfer_status agent_status;
    int id_num = 1;
    unsigned long id_list[1] = {6166};
    char *file_content = "This is file content.\n";
    unsigned long file_len = strlen(file_content);

    rmt_server_send_file(id_list, id_num, "custom_callback", "myfile.txt", file_content, file_len);
    while ((agent_status = rmt_server_get_result(id_list[0], &file_result)) == STATUS_RUNNING) {
        ;
    }
    printf("status: %d, result: %d\n", agent_status, file_result.result);
}

void server_cmd_recv_file(void)
{
    transfer_result file_result;
    transfer_status agent_status;
    unsigned long id = 6166;

    rmt_server_recv_file(id, "custom_callback", "myfile.txt");
    while ((agent_status = rmt_server_get_result(id, &file_result)) == STATUS_RUNNING) {
        ;
    }
    printf("status: %d, result: %d, file_content: %s\n", agent_status, file_result.result, (char *)file_result.pFile);
}

int main(int argc, char *argv[])
{
    int cmd_opt = 0;
    SVR_CMD svr_cmd = CMD_SEARCH;

    // Parse argument
    while ((cmd_opt = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
        switch (cmd_opt) {
            case 'n':
                strcpy(interface, optarg);
                my_interface = interface;
                break;
            case 'h':
                print_help();
                return 0;
            case '?':
            default:
                printf("Not supported option\n");
                return 1;
        }
    }

    if (argc > optind) {
        for (int i = 0; i < CMD_SUM; i++) {
            if (strcmp(svr_cmd_mapping[i], argv[optind]) == 0) {
                svr_cmd = i;
                break;
            }
        }
    }

    printf("RMT Library version is %s\n", rmt_server_version());
    rmt_server_configure(my_interface, 0);
    rmt_server_init();

    switch (svr_cmd) {
        case CMD_SEARCH:
            server_cmd_search_and_get();
            break;
        case CMD_SET:
            server_cmd_set();
            break;
        case CMD_GET_TASK:
            server_cmd_get_task();
            break;
        case CMD_SET_TASK:
            server_cmd_set_task();
            break;
        case CMD_SEND_FILE:
            server_cmd_send_file();
            break;
        case CMD_RECV_FILE:
            server_cmd_recv_file();
            break;
        case CMD_ALL:
            server_cmd_search_and_get();
            server_cmd_set();
            server_cmd_send_file();
            server_cmd_recv_file();
            break;
        default:
            printf("No such function.\n");
            break;
    }

    // free resource
    rmt_server_deinit();

    return 0;
}

File: examples/Python/server_example.py
#!/usr/bin/env python3
import rmt_py_wrapper
import json
import sys, getopt
import time
import psutil
import socket

def usage():
    print("Usage:")
    print("\t-g | --get_config")
    print("\t-s | --set_config")
    print("\t-n eth0 | --net-intf eth0")
    print("\t--send_file")
    print("\t--recv_file")
    print("Example:")
    print("\t./server_example.py -gs -n eth0")

def get_config(dev_list, dev_num):
    r"""
    Get current configuration and status from the given devices

    The following APIs are used to implement this function:
    - rmt_py_wrapper.rmt_server_get_info()
    """

    # Create config key string
    config_list = ["cpu", "ram", "hostname", "wifi"]
    config_key_str = ""
    for item in config_list:
        config_key_str += item + ';'

    # Get device info list
    id_list = rmt_py_wrapper.ulong_array(dev_num)
    for i in range(0, dev_num):
        id_list[i] = dev_list[i].deviceID
    info_num_ptr = rmt_py_wrapper.new_intptr()
    info_list = rmt_py_wrapper.data_info_list.frompointer(rmt_py_wrapper.rmt_server_get_info(id_list, dev_num, config_key_str, info_num_ptr))
    info_num = rmt_py_wrapper.intptr_value(info_num_ptr)
    rmt_py_wrapper.delete_intptr(info_num_ptr) # release info_num_ptr

    print("=== get config result ===")
    config_data = []
    for i in range(0, info_num):
        # Split the result string into dictionary data
        result_list = info_list[i].value_list.split(";")
        dict_data = {"deviceID": info_list[i].deviceID}
        print("deviceID=%d" % info_list[i].deviceID)
        for item in result_list:
            for key in config_list:
                if key in item:
                    dict_data[key] = item[len(key)+1:]
        # print(dict_data)
        config_data.append(dict_data)
    result = json.dumps(config_data, indent=4)
    print(result)

    # Free info_list
    rmt_py_wrapper.rmt_server_free_info(info_list.cast())

    return config_data

def set_diff_config():
    r"""
    Set different configurations or states to the given devices

    The following APIs are used to implement this function:
    - rmt_py_wrapper.rmt_server_set_info()
    """

    # Create data_info_array to set config
    dev_num = 2
    data_info_array = rmt_py_wrapper.new_data_info_array(dev_num)

    # Set for device 5566:
    data_info_element = rmt_py_wrapper.data_info()
    data_info_element.deviceID = 5566
    data_info_element.value_list = "hostname:rqi-1234;locate:on"
    dev_idx = 0
    rmt_py_wrapper.data_info_array_setitem(data_info_array, dev_idx, data_info_element)

    # Set for device 6166:
    data_info_element.deviceID = 6166
    data_info_element.value_list = "hostname:new_hostname;locate:on"
    dev_idx = 1
    rmt_py_wrapper.data_info_array_setitem(data_info_array, dev_idx, data_info_element)

    # Print what we want to set in data_info_array
    print("=== set diff config req ===")
    dev_idx = 0
    data_info_element = rmt_py_wrapper.data_info_array_getitem(data_info_array, dev_idx)
    print("deviceID=%d" % data_info_element.deviceID)
    print("value_list=%s" % data_info_element.value_list)
    dev_idx = 1
    data_info_element = rmt_py_wrapper.data_info_array_getitem(data_info_array, dev_idx)
    print("deviceID=%d" % data_info_element.deviceID)
    print("value_list=%s" % data_info_element.value_list)

    # Send data_info_array to RMT library
    info_num_ptr = rmt_py_wrapper.new_intptr()
    info_list = rmt_py_wrapper.data_info_list.frompointer(rmt_py_wrapper.rmt_server_set_info(data_info_array, dev_num, info_num_ptr))
    info_num = rmt_py_wrapper.intptr_value(info_num_ptr)
    rmt_py_wrapper.delete_intptr(info_num_ptr) # release info_num_ptr

    print("=== set diff config result ===")
    config_data = []
    for i in range(0, info_num):
        # Split the result string into dictionary data
        result_list = info_list[i].value_list.split(";")
        dict_data = {"deviceID": info_list[i].deviceID}
        # print(info_list[i].deviceID)
        # print(info_list[i].value_list)
        for item in result_list:
            key_value_pair = item.split(":")
            if len(key_value_pair) > 1:
                key = key_value_pair[0]
                value = key_value_pair[1]
                dict_data[key] = value
        # print(dict_data)
        config_data.append(dict_data)
    result = json.dumps(config_data, indent=4)
    print(result)

    # Free info_list
    rmt_py_wrapper.rmt_server_free_info(info_list.cast())

def set_same_config():
    r"""
    Set the same configuration or state to each given devices

    The following APIs are used to implement this function:
    - rmt_py_wrapper.rmt_server_set_info_with_same_value()
    """

    # Prepare mock data for setting config
    dev_num = 2
    id_list = rmt_py_wrapper.ulong_array(dev_num)
    id_list[0] = 5566
    id_list[1] = 5567
    config_str = "hostname:rqi-1234;locate:on"

    # Send data_info_array to RMT library
    info_num_ptr = rmt_py_wrapper.new_intptr()
    info_list = rmt_py_wrapper.data_info_list.frompointer(rmt_py_wrapper.rmt_server_set_info_with_same_value(id_list, dev_num, config_str, info_num_ptr))
    info_num = rmt_py_wrapper.intptr_value(info_num_ptr)
    rmt_py_wrapper.delete_intptr(info_num_ptr) # release info_num_ptr

    print("=== set same config result ===")
    config_data = []
    for i in range(0, info_num):
        # Split the result string into dictionary data
        result_list = info_list[i].value_list.split(";")
        dict_data = {"deviceID": info_list[i].deviceID}
        # print(info_list[i].deviceID)
        # print(info_list[i].value_list)
        for item in result_list:
            key_value_pair = item.split(":")
            if len(key_value_pair) > 1:
                key = key_value_pair[0]
                value = key_value_pair[1]
                dict_data[key] = value
        # print(dict_data)
        config_data.append(dict_data)
    result = json.dumps(config_data, indent=4)
    print(result)

    # Free info_list
    rmt_py_wrapper.rmt_server_free_info(info_list.cast())

def discover():
    r"""
    Discover all the available agents in the same network

    The following APIs are used to implement this function:
    - rmt_py_wrapper.rmt_server_create_device_list()
    """

    num_ptr = rmt_py_wrapper.new_intptr()
    dev_list = rmt_py_wrapper.device_info_list.frompointer(rmt_py_wrapper.rmt_server_create_device_list(num_ptr))
    num = rmt_py_wrapper.intptr_value(num_ptr)
    rmt_py_wrapper.delete_intptr(num_ptr) # release num_ptr

    # Put data in JSON format
    data = {"total": num, "items": []}
    items = []
    for i in range(0, num):
        item = {
            "ID": dev_list[i].deviceID,
            "Model": dev_list[i].model,
            "Host": dev_list[i].host,
            "IP": dev_list[i].ip,
            "MAC": dev_list[i].mac,
            "RMT_VERSION": dev_list[i].rmt_version,
            "Device_Info": dev_list[i].devinfo
        }
        items.append(item)

    print("=== discover result ===")
    data["items"] = items
    result = json.dumps(data, indent=4)
    print(result)
    return dev_list, num

def test_send_binary():
    r"""
    Send a binary file to the target device(s)

    The following APIs are used to implement this function:
    - rmt_py_wrapper.rmt_server_send_file()
    - rmt_py_wrapper.rmt_server_get_result()
    """

    print("=== test send binary ===")
    custom_callback = "custom_callback"
    filename = "my_testfile"
    bytes_buffer = b"a\0bc\r\ndef\tg" # convert to bytes
    dev_num = 1
    target_id = 6166
    id_list = rmt_py_wrapper.ulong_array(dev_num)
    id_list[0] = target_id

    agent_status = rmt_py_wrapper.rmt_server_send_file(id_list, dev_num, custom_callback, filename, bytes_buffer)
    print("send_file: agent_status=%d" % agent_status)

    agent_status, result, byte_array = rmt_py_wrapper.rmt_server_get_result(target_id)
    while agent_status == rmt_py_wrapper.STATUS_RUNNING:
        print("sleep for 1 second")
        time.sleep(1)
        agent_status, result, byte_array = rmt_py_wrapper.rmt_server_get_result(target_id)

    print("get_result: agent_status=%d" % agent_status)
    print("transfer_result=%d" % result)
    print(bytes(byte_array).decode("utf-8"))

def test_recv_binary():
    r"""
    Retrieve a binary file from a target device

    The following APIs are used to implement this function:
    - rmt_py_wrapper.rmt_server_recv_file()
    - rmt_py_wrapper.rmt_server_get_result()
    """

    print("=== test recv binary ===")
    target_id = 6166
    custom_callback = "custom_callback"
    filename = "my_testfile"

    agent_status = rmt_py_wrapper.rmt_server_recv_file(target_id, custom_callback, filename)
    print("recv_file: agent_status=%d" % agent_status)

    agent_status, result, byte_array = rmt_py_wrapper.rmt_server_get_result(target_id)
    while agent_status == rmt_py_wrapper.STATUS_RUNNING:
        print("sleep for 1 second")
        time.sleep(1)
        agent_status, result, byte_array = rmt_py_wrapper.rmt_server_get_result(target_id)

    print("get_result: agent_status=%d" % agent_status)
    print("transfer_result=%d" % result)
    print("file_len=%d" % len(byte_array))
    print("=== file content start ===")
    print(bytes(byte_array).decode("utf-8"))
    print("=== file content end ===")

def main(args):
    r"""
    Init and de-init RMT server library

    The following APIs are used to implement this function:
    - rmt_py_wrapper.rmt_server_version()
    - rmt_py_wrapper.rmt_server_configure()
    - rmt_py_wrapper.rmt_server_init()
    - rmt_py_wrapper.rmt_server_deinit()
    """

    def valid_interface(interface):
        interface_addrs = psutil.net_if_addrs().get(interface) or []
        return socket.AF_INET in [snicaddr.family for snicaddr in interface_addrs]

    try:
        opts, args = getopt.getopt(args, "hgsn:", ["help", "get_config", "set_config", "send_file", "recv_file", "net_intf="])
    except getopt.GetoptError as err:
        # print help information and exit:
        print(err)  # will print something like "option -a not recognized"
        usage()
        sys.exit(2)
    flag_get_config = False
    flag_set_config = False
    flag_send_file = False
    flag_recv_file = False
    my_interface = ""
    for opt, arg in opts:
        if opt in ("-h", "--help"):
            usage()
            sys.exit()
        elif opt in ("-g", "--get_config"):
            flag_get_config = True
        elif opt in ("-s", "--set_config"):
            flag_set_config = True
        elif opt in ("--send_file"):
            flag_send_file = True
        elif opt in ("--recv_file"):
            flag_recv_file = True
        elif opt in ("-n", "--net-intf"):
            my_interface = arg
            if not valid_interface(my_interface):
                print("Interface({}) is invalid or inactive.".format(my_interface))
                sys.exit(2)
        else:
            assert False, "unhandled option"

    # Get RMT_VERSION
    print("RMT_VERSION=%s" % rmt_py_wrapper.rmt_server_version())

    # Set network interface for DDS communication
    print("Use interface({}) for RMT server".format(my_interface))
    rmt_py_wrapper.rmt_server_configure(my_interface, 0)

    # Init RMT server
    rmt_py_wrapper.rmt_server_init()

    # Discovery devices
    dev_list, num = discover()

    # Get config
    if flag_get_config:
        get_config(dev_list, num)

    # Set config
    if flag_set_config:
        set_same_config()
        set_diff_config()

    # Send file
    if flag_send_file:
        test_send_binary()

    # Recv file
    if flag_recv_file:
        test_recv_binary()

    # Free & de-init
    rmt_py_wrapper.rmt_server_free_device_list(dev_list.cast())
    rmt_py_wrapper.rmt_server_deinit()

if __name__ == "__main__":
    args = sys.argv[1:]
    main(args)
